### PR TITLE
feature: Add Free U

### PR DESCRIPTION
**Pull request by isamu-isozaki**
Hi! This PR is a draft. FreeU's main idea came from the finding that the mid block of the UNet is responsible for denoising and the high frequency textures come from the skip connections. 
The denoising strength can be improved by scaling half of the channels from the mid block by a factor
The high frequency detail can be improved by lowering the low frequency features in the skip connection.

My current plan is to add FreeU to this library to the first 2 upsample blocks in the UNet with an adapter. The main strategy is to replace the ResidualConcatenator for the first 2 upsample block res blocks and then do the FreeU operations then.
Once that is done I think the idea is done.  I'm new to this library so happy for suggestions!

### CODE REVIEW

**Comment by isamu-isozaki** [comment]
Ok! I think I am done with the initial idea. Will test now

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    FreeUSkipConnections(s)
+                ),
+                dim=1,
+            ),
+        )
+
+class SD1UnetFreeUAdapter:
+    def __init__(self, unet, bs: [float], ss: [float]):
+        assert len(bs) == len(ss)
+        self.unet = unet
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:74]: For composability reasons, we always realise adapters as a fl.Chain with the Adapter class.

You can take example on the SelfAttentionGuidance which is also a weightless adapter on top of SDUNet: https://github.com/finegrain-ai/refiners/blob/main/src/refiners/foundationals/latent_diffusion/self_attention_guidance.py#L67

Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:74]: Thanks! Fixing now.

**Comment by isamu-isozaki** [comment]
Some test results with the below script:
```
import torch

from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.foundationals.latent_diffusion.freeu import SD1UnetFreeUAdapter
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
sd1 = StableDiffusion_1(device="cuda", dtype=torch.float16)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
free_u_adapter = SD1UnetFreeUAdapter(sd1.unet, bs=[1.2, 1.4], ss=[0.9, 0.2])
def test_prompt(prompt, output_file):
    seed = 9752
    first_step = 1
    sd1.set_num_inference_steps(50)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"best quality, high quality {prompt}", negative_text="monochrome, lowres, bad anatomy, worst quality, low quality"
        )
    
        manual_seed(seed=seed)
        x = sd1.init_latents(size=(512, 512), first_step=first_step).to(
            device=sd1.device, dtype=sd1.dtype
        )
    
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding,
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
os.makedirs("freeu", exist_ok=True)
os.makedirs("no_freeu", exist_ok=True)
free_u_adapter.inject()
test_prompt("cute dog", "freeu/output_dog.jpg")
free_u_adapter.eject()
test_prompt("cute dog", "no_freeu/output_dog.jpg")
```
gave
output_dog
for freeU
and
output_dog
for not freeu

File: src/refiners/foundationals/latent_diffusion/freeu.py (hunk 1 above)
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py]: I think fixed. Let me know what you think

**Comment by isamu-isozaki** [comment]
@limiteinductive I think I did everything I can think of for now. So should be ready for review!

**Comment by limiteinductive** [comment]
I can reproduce your result for SD1.5 allright, well played! I tried SDXL and for this script I get these results
`    import torch
    from pathlib import Path
    from refiners.foundationals.latent_diffusion.stable_diffusion_xl import StableDiffusion_XL
    from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
    from refiners.fluxion.utils import manual_seed

    sd = StableDiffusion_XL(device="cuda", dtype=torch.float32)
    model_path = Path("/mnt/ssd2/hub/finegrain/stable-diffusion-xl/")
    sd.lda.load_from_safetensors(model_path / "sdxl-lda.safetensors")
    sd.unet.load_from_safetensors(model_path / "sdxl-unet.safetensors")
    sd.clip_text_encoder.load_from_safetensors(model_path / "DoubleCLIPTextEncoder.safetensors")
    free_u_adapter = SDFreeUAdapter(sd.unet, bs=[1.1, 1.2], ss=[0.6, 0.4])

    def test_prompt(prompt: str, output_file: Path | str) -> None:
        seed = 9752
        first_step = 1
        sd.set_num_inference_steps(50)
        with torch.no_grad():
            clip_text_embedding, pooled_text_embedding = sd.compute_clip_text_embedding(
                text=f"best quality, high quality {prompt}",
                negative_text="monochrome, lowres, bad anatomy, worst quality, low quality",
            )
            time_ids = sd.default_time_ids

            manual_seed(seed=seed)
            x = sd.init_latents(size=(1024, 1024), first_step=first_step).to(device=sd.device, dtype=sd.dtype)

            for step in sd.steps[first_step:]:
                x = sd(
                    x,
                    step=step,
                    clip_text_embedding=clip_text_embedding,
                    pooled_text_embedding=pooled_text_embedding,
                    time_ids=time_ids,
                )
            predicted_image = sd.lda.decode_latents(x=x)

        predicted_image.save(output_file)
        print(f"done: see {output_file}")

    Path("freeu").mkdir(exist_ok=True)
    Path("no_freeu").mkdir(exist_ok=True)
    free_u_adapter.inject()
    test_prompt("cute dog", "freeu/output_dog.jpg")
    free_u_adapter.eject()
    test_prompt("cute dog", "no_freeu/output_dog.jpg")`
    
    no free_u:
    
image

with free_u:
image

I tried to play a bit with the parameters, but it doesn't seem to fix the issue

**Comment by isamu-isozaki** [comment]
Interesting. My current guess is that there's too much high-frequency noise coming through. I'm thinking of testing lowering the s parameters more.

**Comment by isamu-isozaki** [comment]
Ok. I think I found the inconsistency between diffusers and my current implementation.
Currently, I apply freeU once before bringing into the top 2 bottlenecks. However, in diffuser's case they do
```
        for resnet in self.resnets:
            # pop res hidden states
            res_hidden_states = res_hidden_states_tuple[-1]
            res_hidden_states_tuple = res_hidden_states_tuple[:-1]

            # FreeU: Only operate on the first two stages
            if is_freeu_enabled:
                hidden_states, res_hidden_states = apply_freeu(
                    self.resolution_idx,
                    hidden_states,
                    res_hidden_states,
                    s1=self.s1,
                    s2=self.s2,
                    b1=self.b1,
                    b2=self.b2,
                )
```
for each resolution index. So tldr diffusers is applying freeU to more blocks if I understand correctly. So I should figure out which blocks get freeU applied and just duplicate some values in ss and bs and I think it will work

**Comment by isamu-isozaki** [comment]
Hi. I think I figured it out. My code is correct for this pr but not for the example code, if you change from 
```
free_u_adapter = SDFreeUAdapter(sd1.unet, bs=[1.2, 1.4], ss=[0.9, 0.2])
```
to
```
free_u_adapter = SDFreeUAdapter(sd1.unet, bs=[1.2, 1.2, 1.2, 1.4, 1.4, 1.4], ss=[0.9, 0.9, 0.9, 0.2, 0.2, 0.2])
```

**Comment by isamu-isozaki** [comment]
In diffusers,

I did notice there are slight differences but then the output for
```
from diffusers import DiffusionPipeline, DPMSolverMultistepScheduler
import torch
from PIL import Image
import os
from refiners.fluxion.utils import manual_seed
sd_with_free_u = "freeu_sd"
sd_with_no_free_u = "no_freeu_sd"
os.makedirs(sd_with_free_u, exist_ok=True)
os.makedirs(sd_with_no_free_u, exist_ok=True)

pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, safety_checker=None
).to("cuda")
pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)
def test_prompt(prompt, output_file):
    seed=9752
    manual_seed(seed=seed)
    image = pipeline(width=512, height=512, prompt=f"best quality, high quality {prompt}", negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality").images[0]
    image.save(output_file)
    print(f"done: see {output_file}")
pipeline.enable_freeu(s1=0.9, s2=0.2, b1=1.2, b2=1.4)
test_prompt("cute cat", f"{sd_with_free_u}/output_cat.jpg")
pipeline.disable_freeu()
test_prompt("cute cat", f"{sd_with_no_free_u}/output_cat.jpg")
```
gives 

no free u:
output_cat

with free u:
output_cat

**Comment by isamu-isozaki** [comment]
in refiners:
```
import torch

from PIL import Image
import os
from refiners.foundationals.latent_diffusion.stable_diffusion_1 import StableDiffusion_1
from refiners.foundationals.latent_diffusion.freeu import SDFreeUAdapter
from refiners.fluxion.utils import manual_seed, image_to_tensor, load_from_safetensors
sd1 = StableDiffusion_1(device="cuda", dtype=torch.float16)
model_path = r"..\refiners\tests\weights"
sd1.lda.load_from_safetensors(os.path.join(model_path, "lda.safetensors"))
sd1.unet.load_from_safetensors(os.path.join(model_path, "unet.safetensors"))
sd1.clip_text_encoder.load_from_safetensors(os.path.join(model_path, "CLIPTextEncoderL.safetensors"))
free_u_adapter = SDFreeUAdapter(sd1.unet, bs=[1.2, 1.2, 1.2, 1.4, 1.4, 1.4], ss=[0.9, 0.9, 0.9, 0.2, 0.2, 0.2])
def test_prompt(prompt, output_file):
    seed = 9752
    first_step = 1
    sd1.set_num_inference_steps(50)
    with torch.no_grad():
        # Note: default text prompts for IP-Adapter
        clip_text_embedding = sd1.compute_clip_text_embedding(
            text=f"best quality, high quality {prompt}", negative_text="monochrome, lowres, bad anatomy, worst quality, low quality"
        )
    
        manual_seed(seed=seed)
        x = sd1.init_latents(size=(512, 512), first_step=first_step).to(
            device=sd1.device, dtype=sd1.dtype
        )
    
        for step in sd1.steps[first_step:]:
            x = sd1(
                x,
                step=step,
                clip_text_embedding=clip_text_embedding,
            )
        predicted_image = sd1.lda.decode_latents(x=x)
    
    predicted_image.save(output_file)
    print(f"done: see {output_file}")
os.makedirs("freeu", exist_ok=True)
os.makedirs("no_freeu", exist_ok=True)
free_u_adapter.inject()
test_prompt("cute cat", "freeu/output_cat.jpg")
free_u_adapter.eject()
test_prompt("cute cat", "no_freeu/output_cat.jpg")
```
gives no free u
output_cat

and free u
output_cat

**Comment by isamu-isozaki** [comment]
there's sometimes some minor differences even when no free u is applied between diffusers and refiners like
output_bird
for sd and
output_bird
for refiners. so the outputs are sometimes a bit different for free u too but usually similar

**Comment by catwell** [comment]
@isamu-isozaki I don't know if that's what you're seeing here, but a difference between Diffusers and Refiners is that in the VAE Diffusers samples from the distribution (`mean + rand * std`) whereas we only take the mean. The std is typically very small and this does not make a big difference by itself, but it means Diffusers changes the state of the random generator which can result in larger differences sometimes.

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    # Non-power of 2 images must be float32
+    if (W & (W - 1)) != 0 or (H & (H - 1)) != 0:
+        x = x.to(dtype=torch.float32)
+
+    # FFT
+    x_freq = fftn(x, dim=(-2, -1))  # type: ignore
+    x_freq = fftshift(x_freq, dim=(-2, -1))  # type: ignore
+
+    B, C, H, W = x_freq.shape  # type: ignore
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:36]: x_freq.shape will always be x.shape right? So this line doesn't seem useful

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    mask = torch.ones((B, C, H, W), device=x.device)  # type: ignore
+
+    crow, ccol = H // 2, W // 2  # type: ignore
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:39]:
```suggestion
    center_row, center_col = height // 2, width // 2  # type: ignore
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    mask[..., crow - threshold : crow + threshold, ccol - threshold : ccol + threshold] = scale
+    x_freq = x_freq * mask  # type: ignore
+
+    # IFFT
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:43]: not needed

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        super().__init__()
+        self.b = b
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        num_half_channels = hidden_states.shape[1] // 2
+        hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b
+        return hidden_states
+
+
+class FreeUSkipConnections(Module):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:62]:
```suggestion
class FreeUSkipConnections(fl.Module):
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, s: float):
+        super().__init__()
+        self.s = s
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:65]:
```suggestion
        self.scale = scale
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:67]:
```suggestion
    def forward(self, x: Tensor) -> Tensor:
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        hidden_states = fourier_filter(hidden_states, threshold=1, scale=self.s)
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:68]: You could even realise FreeUSkipConnections
```suggestion
        return fourier_filter(x, scale=self.scale)
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        return hidden_states
+
+
+class FreeUResidualConcatenator(fl.Chain):
+    def __init__(self, n: int, b: float, s: float) -> None:
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:73]: (feel free to suggest other names)
```suggestion
    def __init__(self, n: int, backbone_scale: float = 1, residual_scale: float = 1) -> None:
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        self.n = n
+
+        super().__init__(
+            fl.Concatenate(
+                FreeUHiddenStates(b),
+                fl.Chain(
+                    fl.UseContext(context="unet", key="residuals").compose(lambda residuals: residuals[self.n]),
+                    FreeUSkipConnections(s),
+                ),
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:82]: Instead of making, you could create a FreeUResidualFilter that inherits from fl.Chain and has two layers

fl.UseContext(context="unet", key="residuals").compose(lambda residuals: residuals[self.n]),
fl.Lambda(lambda x: fourier_filter(x, scale=self.scale))

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, s: float):
+        super().__init__()
+        self.s = s
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        hidden_states = fourier_filter(hidden_states, threshold=1, scale=self.s)
+        return hidden_states
+
+
+class FreeUResidualConcatenator(fl.Chain):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:72]: you can directly inherit from fl.Concatenate

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+class FreeUHiddenStates(Module):
+    def __init__(self, b: float):
+        super().__init__()
+        self.b = b
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        num_half_channels = hidden_states.shape[1] // 2
+        hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b
+        return hidden_states
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:59]: dito

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+    # IFFT
+    x_freq = ifftshift(x_freq, dim=(-2, -1))  # type: ignore
+    x_filtered = ifftn(x_freq, dim=(-2, -1)).real  # type: ignore
+
+    return x_filtered.to(dtype=x_in.dtype)  # type: ignore
+
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+class FreeUHiddenStates(Module):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:51]: Try to find a more explicit name for this module

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                    fl.UseContext(context="unet", key="residuals").compose(lambda residuals: residuals[self.n]),
+                    FreeUSkipConnections(s),
+                ),
+                dim=1,
+            ),
+        )
+
+
+class SDFreeUAdapter(Generic[T], fl.Chain, Adapter[T]):
+    def __init__(self, target: T, bs: list[float], ss: list[float]):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:89]: try to find explicit names for bs and ss

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        assert len(bs) == len(ss)
+        self.bs = bs
+        self.ss = ss
+        with self.setup_adapter(target):
+            super().__init__(target)
+
+    def inject(self: TSDFreeUAdapter, parent: fl.Chain | None = None) -> TSDFreeUAdapter:
+        for n, block in enumerate(cast(Iterable[fl.Chain], self.target.UpBlocks)):
+            if n >= len(self.bs):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:98]: instead of iteratiing through the blocks, iterate on enumerate(zip(bs, ss)) so you don't need a break

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+                break
+            block.pop(0)
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:100]: instead of pop insert, you can use the replace method

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+            block.insert(0, FreeUResidualConcatenator(-n - 2, self.bs[n], self.ss[n]))
+        return super().inject(parent)
+
+    def eject(self) -> None:
+        for n, block in enumerate(cast(Iterable[fl.Chain], self.target.UpBlocks)):
+            if n >= len(self.bs):
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:106]: dito

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+                break
+            block.pop(0)
```
Comment (limiteinductive) [inline src/refiners/foundationals/latent_diffusion/freeu.py:108]: dito

**Review by limiteinductive — CHANGES_REQUESTED** [review]
Essentially just some nits and cleaning bits, wp!

File: src/refiners/foundationals/latent_diffusion/freeu.py (hunk 10 above)
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:82]: wow that's pretty awesome. I think I did make it more readable by replacing the FreeUSkipConnections to be that

**Comment by isamu-isozaki** [comment]
@limiteinductive Thanks for the review! I did rename a bit. The reason I named it skip_scales and backbone_scales is because in the paper
image
the b applies to backbone features and s applies to skip features. Let me know what you think!

File: src/refiners/foundationals/latent_diffusion/freeu.py (hunk 9 above)
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py]: yup! Mentioned below but maybe backbone_scale feels great but for the other I think skip_scale is better to make it consistent with paper

File: src/refiners/foundationals/latent_diffusion/freeu.py (hunk 12 above)
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py]: I made the hidden_states x to keep it consistent

**Comment by isamu-isozaki** [comment]
I tested the current change and produced same output

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        assert len(backbone_scales) == len(skip_scales)
+        self.backbone_scales = backbone_scales
+        self.skip_scales = skip_scales
+        with self.setup_adapter(target):
+            super().__init__(target)
+
+    def inject(self: TSDFreeUAdapter, parent: fl.Chain | None = None) -> TSDFreeUAdapter:
+        for n, (backbone_scale, skip_scale) in enumerate(zip(self.backbone_scales, self.skip_scales)):
+            block = self.target.UpBlocks[n]
+            block.replace(block[0], FreeUResidualConcatenator(-n - 2, backbone_scale, skip_scale))
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/freeu.py:85]:
```suggestion
            concat = block.ensure_find(ResidualConcatenator)
            block.replace(concat, FreeUResidualConcatenator(-n - 2, backbone_scale, skip_scale))
```

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+        return super().inject(parent)
+
+    def eject(self) -> None:
+        for n in range(len(self.backbone_scales)):
+            block = self.target.UpBlocks[n]
+            block.replace(block[0], ResidualConcatenator(-n - 2))
```
Comment (deltheil) [inline src/refiners/foundationals/latent_diffusion/freeu.py:91]:
```suggestion
            concat = block.ensure_find(FreeUResidualConcatenator)
            block.replace(concat, ResidualConcatenator(-n - 2))
```

**Review by deltheil — COMMENTED** [review]
👏 super nice PR! Sent a couple of nits

File: src/refiners/foundationals/latent_diffusion/freeu.py
[the remark this answers is not in the document]
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:1]: oh weird I thought I ran it. Let me try again!

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    x_freq = x_freq * mask  # type: ignore
+
+    x_freq = ifftshift(x_freq, dim=(-2, -1))  # type: ignore
+    x_filtered = ifftn(x_freq, dim=(-2, -1)).real  # type: ignore
+
+    return x_filtered.to(dtype=x.dtype)  # type: ignore
+
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+class FreeUIdentity(fl.Module):
```
[the remark this answers is not in the document]
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:46]: ah yup that's way better!

File: src/refiners/foundationals/latent_diffusion/freeu.py [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+def fourier_filter(x: Tensor, scale: float = 1, threshold: int = 1) -> Tensor:
+    """Fourier filter as introduced in FreeU (https://arxiv.org/abs/2309.11497).
+
+    This version of the method comes from here:
+    https://github.com/ChenyangSi/FreeU/blob/main/demo/free_lunch_utils.py#L23
+    """
+    batch, channels, height, width = x.shape
```
[the remark this answers is not in the document]
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:22]: I did the same for x.device to be consistent

File: src/refiners/foundationals/latent_diffusion/freeu.py
[the remark this answers is not in the document]
Comment (isamu-isozaki) [reply src/refiners/foundationals/latent_diffusion/freeu.py:1]: Just ran it

**Comment by isamu-isozaki** [comment]
@deltheil I think I did it! Let me know what you think

**Comment by deltheil** [comment]
> @deltheil I think I did it! Let me know what you think

LGTM. Last thing: could you please squash your commits into a single one? And then we are good to go. Thanks!

**Comment by isamu-isozaki** [comment]
@deltheil Thanks! And let me know if I did it correctly. I'm still not too familiar with squashing commits.

**Comment by deltheil** [comment]
@isamu-isozaki looks good after your last push, thanks!

**Review by deltheil — APPROVED** [review]

**Comment by isamu-isozaki** [comment]
Thanks!
